### DISCUSSION TITLE
Using telemetrix (arduino or esp32) with motor shields.

### DISCUSSION THREAD

**Discussion by sygolene** [MrYsLab/telemetrix#67] [Q&A] [answered]
Hello, 

I've been using firmata for a while using my own prototype circuit using a arduino nano + A4988 stepper motor driver and have recently decided to move away from "own circuits" to transition to arduino & motor shields to avoid debugging my poor circuit design and have something more reliable, as well as to transition to telemetrix which I like more (documentation, features, good support from author)
Pardon the question if it is naive, but I thought I could use telemetrix with these shields without having to write custom code by simply adressing the sda & scl ports?

the stack is as follows : 
-arduino UNO R4 wifi (used with Telemetrix4arduino.ino as I don't need wifi)
-adafruit motorshield v2.3
-python telemetrix library 

if my research is correct,
STEP & DIR from the A4988 I was using are simply SDA & SCL respectively ,and these on the UNO R4 would be A4 (pin 18) & A5 (pin 19), respectively. 
So in my python code, I'd map accordingly: 
PULSE_PIN = 18 
DIRECTION_PIN = 19 
yet it doesn't seem to work. It is my understanding unless you start stacking, the I/O are "normally accessible" ? 
I have debugged the circuit using pure arduino code with the adafruit shield libraries and the motor is moving accordingly. 

Thanks in advance,

**Comment by sygolene** [comment]
Oh I think I just know why it doesn't work... the shield is normally driven by specific code from its driver library...
So just sending data on sda and scl doesn't make any sense :'(... 
I suppose the solution would be to override your stepper functions with the ones from adafruit libraries, or to wrap them inside or something like that...hummm perhaps too difficult with my existing skills.. 
Or do you think it would be a good idea to support a couple of typical motor shields?
Any feedback welcome :-)

**Comment by MrYsLab** [comment]
Hi,
Thanks for the question. Are you using an A4988 or the Adafruit Motorshield v2.3? Before we get to that, for the UNO R4, the Python library you should use is telemetrix-uno-r4. You will also need to install the matching Arduino library, Telemetrix4UnoR4. The library offers a choice of three transports: WiFi, USBSerial, and BLE. From your email, it sounds like you wish to use the USBSerial version.

The original intent for this library was to include an AccelStepper implementation. However, AccelStepper is not currently compatible with the Uno R4. I reported this to the AccelStepper author, but the problem still needs to be corrected.

If you are using the Nano, then telemetrix and Telemetrix4Arduino should work. The telemetrix library has the accelstepper interface included. However, I only tested using a TB6600 driver and not the A4988.

For the motor drivers, I don't believe either use I2C when controlling stepper motors. For the Adafruit board, I believe I2C is for controlling servo motors and not steppers. Therefore, SDA and SCL are not used for control.

If you still need help, please let me know if you are using a UNO R4 or the Nano and if you are using it in conjunction with the A4988 or Adafruit driver.

Thanks,
Alan

**Comment by MrYsLab** [comment]
I just saw your second comment. For the Adafruit board, the documentation states that it should work with the AccelStepper library. I am unsure if that means that you may not need to use their Arduino library at all.

My implementation of AccelStepper may support both the Adafruit board and the A4988. Again, for the UNO R4, stepper motors are not supported at this time.

Here is a [telemetrix stepper example](https://github.com/MrYsLab/telemetrix/blob/master/examples/stepper_absolute.py).

**Comment by sygolene** [comment]
Hello Alan, 

Thanks very much for your first replies, I'll have a look in more detail soon. I thought I'd clarify a bit my path.

I can already confirm that telemetrix was indeed working on my "own" build which was an arduino nano + A4988, at least intermittently. It's the fact it wasnt' working consistently that got me looking at ready made circuits to isolate my poor craftman skills from the debugging process. 
That's when I decided to transition from nano + A4988 to two alternatives:
-Arduino R4 + Adafruit motorshield v2.3 (TB6612 driver) https://learn.adafruit.com/adafruit-motor-shield-v2-for-arduino
-Adafruit feather ESP32-S2 Reverse TFT + Adafruit featherwing motorshield (TB6612) https://learn.adafruit.com/adafruit-stepper-dc-motor-featherwing - here I haven't started to look at it. The idea with this one is that I'd like to have the LCD working too, possibly overloading your library, but I'm not there yet.
In both case, they use TB6612 and the same adafruit libraries 
```
#include <Wire.h>
#include <Adafruit_MotorShield.h>
#include "utility/Adafruit_PWMServoDriver.h"
```
Regarding "For the Adafruit board, I believe I2C is for controlling servo motors and not steppers. Therefore, SDA and SCL are not used for control.", note in the page of the featherwing the following passage "Since the FeatherWing only uses the I2C (SDA & SCL pins), it works with any and all Feathers. ". Also, on their older page for their V1 library (deprecated), they mention " V2 of the shield uses i2c only and works with anything that has I2C support (e.g. all arduinos) without endless incompatibilities and porting requirements". https://adafruit.github.io/Adafruit_Motor_Shield_V2_Library/html/index.html

My first goal now, is to have a nice simple way to have the stepper motor work using telemetrix with these shields. 
I'm still under the impression I'll have to fiddle with their code & yours one way or another as you didn't write code to control the TB6612?
I'm still relatively novice to all this I'm afraid! 
I'll first try using the correct libraries and will get back to you :-) thanks for your patience,

**Comment by MrYsLab** [comment]
Thanks for the clarification. To reiterate, the UNO R4 currently does not support the AccelStepper library.

For the Nano, you may be able to use the Adafruit Motor Shield v2 without any changes to telemetrix or Telemetrix4Arduino. However, you will need to figure out the direction and pulse pin the board uses for each motor. I looked at the Adafruit code, and it is very poorly commented, so you could contact Adafruit and ask them what PINs they are using.

For the LCD display, I do not understand how that board connects to the motor shield and the software that drives it. To use that board with telemetrix, assuming it uses an i2c interface, you would need to port the Adafruit i2c code to telemetrix Python.

I do have support for simpler i2c LCD display. You may find the code [here](https://github.com/MrYsLab/telemetrix/tree/master/examples/Contributed/i2c_lcd_pcf8574).

**Comment by MrYsLab** [comment]
I quickly looked at the Adafruit feather ESP32-S2 Reverse TFT and could not find its Arduino source code. The control for this board is accomplished with CircuitPython. Using this board with telemetrix may be extremely difficult, if not impossible.

If I can help in any way, please let me know.

**Comment by sygolene** [comment]
Thanks.!

I still have that nano and an R3
I gave a quick try with the R3, I did try reading through
https://github.com/adafruit/Adafruit_Motor_Shield_V2_Library/blob/master/Adafruit_MotorShield.cpp#L178
and tried to map to DIR & STEP  as "bypass" hoping Telemetrix4arduino would
then behave natively and carry over these to AccelStepper but no luck.

Frankly, by now it's just pure random trials. Adafruit wrote code to handle
that TB6612 in its lib and doesn't seem to be using AccelStepper at all so
just sending data over "some" pin is probably not going to cut it I'm
afraid :'(. Besides, I don't even know if the shield's pins are accessed
like this.

About the ESP32 S2 reverse, I thought it was either running micropython or
arduino. It's just a matter of bootloader. I tried both and use it
currently plainly in the arduino IDE.

So I initially intended to use the sexy ESP32 S2 with tft display with a
telemetrix I'd overload with LCD but was willing to start "simple" with a
plain UNO R4. Now neither R4 or this ESP32 board are fully compatible with
either telemetrix or accelstepper... hmmm

So... my options...
a- either I go deeper in the
https://github.com/adafruit/Adafruit_Motor_Shield_V2_Library/blob/master/Adafruit_MotorShield.cpp
understanding and port some of it to work for R4, then going LCD later with
R4
b- back to nano & A4988 , my initial circuit, hoping I can do some
c- give a try to another "native" config with a TB6612 chip
I fear A is a dangerous part to non standard non maintainable code and too
little experience  so b or c..
What do you think ?

**Comment by sygolene** [comment]
There was option  "d", implement serial comms between client and server
myself but seemed dumb when proper libs exist ;)..

**Comment by MrYsLab** [comment, marked as answer]
Looking at the Adafruit documentation in more detail, I am not qualified to advise you on using the motor shield with telemetrix. I cannot vouch for [this article](https://hackaday.io/project/183279-accelstepper-the-missing-manual/details), but it states:

> The following interfaces are ones I have no experience with but are supported by the library.
AccelStepper::FULL2WIRE (2 wire stepper, 2 motor pins required.)
AccelStepper::FULL3WIRE (3 wire stepper, such as HDD spindle, 3 motor pins required.)
AccelStepper::HALF3WIRE (3 wire half stepper, such as HDD spindle, 3 motor pins required.)
AccelStepper::HALF4WIRE (4 wire half stepper, 4 motor pins required.)
The last interface is used primarily by Adafruit for their motor interface board. It is best documented by Adafruit and I will not discuss it further.
AccelStepper::FUNCTION (Use the functional interface, implementing your own driver functions, as Adafruit does.)

Essentially, Adafruit developed its own custom interface. If you could reverse engineer their code, adding it to Telemetrx4Arduino might not be worth it. However, I can advise you on what needs to be done from the point of view of the Telemetrix4Arduino architecture, but I would be of little help with the actual driving code.

The only stepper driver that I have tested with telemetrix is [this one](https://www.amazon.com/gp/product/B07HHS14VQ/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&th=1).

I found this [article on interfacing the A4988 to the Arduino](https://www.makerguides.com/a4988-stepper-motor-driver-arduino-tutorial/). It includes a section on using AccelStepper. If you can reproduce the author's results, then it might be worthwhile trying telemetrix with that setup.

As far as the LCD, I don't see a way to control it with telemetrix since the display sits on top of an ESP32.
There is an [ESP32 version of telemetrix](https://github.com/MrYsLab/Telemetrix4Esp32), but it supports WIFI and BLE and not SerialUSB.

Please let me know if I can help in any way with your project.

**Reply by sygolene** [reply]
Hello Allan, 

Indeed, Adafruit did their full implementation as the cpp shows. 
Thanks for the link on the A4988, it's actually that very schema I followed and can confirm worked with telemetrix & firmata.
I'm currently fiddling back with option d and playing with serial communications but I'm pretty sure I'll bump into a lot of surprises and probably won't match your level or error control protocol. But perhaps is it not needed for my current project. 

I'd like to thank you for your replies, appreciated !
